osquery/core/database/rocksdb_migration.cpp

```diff
@@ -0,0 +1,228 @@
+      return version_result.takeError();
+    }
+  } else {
+    // InvalidArgument means that db does not exists
+    if (open_result.getErrorCode() == RocksdbMigrationError::InvalidArgument) {
+      return Success();
+    } else {
+      return open_result.takeError();
+    }
+  }
+  return Success();
+}
+
+ExpectedSuccess<RocksdbMigrationError> RocksdbMigration::dropDbMigration(
+    const std::string& src_path, const std::string& dst_path) {
+  auto dst_db = openDatabase(dst_path, true, true);
+  if (dst_db) {
+    return Success();
+  }
+  return dst_db.takeError();
+}
+
+void RocksdbMigration::buildMigrationMap() {
+  // No migration map at this stage, will be added as separate commit
+}
+
+ExpectedSuccess<RocksdbMigrationError> RocksdbMigration::migrateFromVersion(
+    int original_version) {
+  buildMigrationMap();
+  std::string src_path = source_path_;
+  std::string dst_path = randomOutputPath();
+  bool drop_src_data = false;
+  for (int version = original_version;
+       version < kDatabaseSchemaVersionCurrent;) {
+    auto iter = migration_map_.find(version);
+    if (iter == migration_map_.end()) {
+      return createError(RocksdbMigrationError::NoMigrationFromCurrentVersion,
+                         "No migration logic from version: ")
+             << version;
+    }
+    VLOG(1) << "Migrating from version: " << version
+            << ". Src path: " << src_path << ". Dst path: " << dst_path;
+    auto migration_result = iter->second(src_path, dst_path);
+    if (migration_result) {
+      int new_version = migration_result.take();
+      if (new_version <= version) {
+        return createError(RocksdbMigrationError::MigrationLogicError,
+                           "New version(")
+               << version << ") is lower or same as old(" << new_version << ")";
+      }
+      version = new_version;
+    } else {
+      return migration_result.takeError();
+    }
+    if (drop_src_data) {
+      VLOG(1) << "Destroying db at path: " << src_path;
+      rocksdb::DestroyDB(src_path, rocksdb::Options());
+    }
+    drop_src_data = true;
+    src_path = dst_path;
+    dst_path = randomOutputPath();
+  }
+
+  std::string temp_store = randomOutputPath();
+  auto original_src = source_path_;
+  auto new_src = src_path;
+
+  auto move1 = moveDb(original_src, temp_store);
+  if (move1) {
+    auto move2 = moveDb(new_src, original_src);
+    if (move2) {
+      rocksdb::DestroyDB(temp_store, rocksdb::Options());
+      return Success();
+    }
+    return move2.takeError();
+  }
+  return move1.takeError();
+}
+
+ExpectedSuccess<RocksdbMigrationError> RocksdbMigration::moveDb(
+    const std::string& src_path, const std::string& dst_path) {
+  if (BOOST_UNLIKELY(fs::exists(fs::path(dst_path)))) {
+    return createError(RocksdbMigrationError::FailMoveDatabase,
+                       "Database at dst path already exists: ")
+           << dst_path;
+  }
+  boost::system::error_code ec;
+  fs::rename(src_path, dst_path, ec);
```

Comment:
I would for using STL https://en.cppreference.com/w/cpp/io/c/rename , but it does not matter much